pallets/registrar/src/tests.rs

```diff
@@ -559,28 +666,344 @@ fn deregister_returns_bond() {
+fn register_deregister_register_in_same_block() {
+    new_test_ext().execute_with(|| {
+        run_to_block(1);
+        assert_ok!(ParaRegistrar::register(
+            RuntimeOrigin::signed(ALICE),
+            42.into(),
+            empty_genesis_data()
+        ));
+        assert_eq!(
+            ParaRegistrar::para_genesis_data(ParaId::from(42)).as_ref(),
+            Some(&empty_genesis_data())
+        );
+        assert_ok!(ParaRegistrar::deregister(RuntimeOrigin::root(), 42.into(),));
+        assert_eq!(
+            ParaRegistrar::para_genesis_data(ParaId::from(42)).as_ref(),
+            None
+        );
+        let new_genesis_data = ContainerChainGenesisData {
+            storage: vec![(b"key".to_vec(), b"value".to_vec()).into()],
+            name: Default::default(),
+            id: Default::default(),
+            fork_id: Default::default(),
+            extensions: Default::default(),
+            properties: Default::default(),
+        };
+        assert_ok!(ParaRegistrar::register(
+            RuntimeOrigin::signed(ALICE),
+            42.into(),
+            new_genesis_data.clone(),
+        ));
+        assert_eq!(
+            ParaRegistrar::para_genesis_data(ParaId::from(42)).as_ref(),
+            Some(&new_genesis_data)
+        );
+        run_to_session(2);
+        assert_eq!(
+            ParaRegistrar::para_genesis_data(ParaId::from(42)).as_ref(),
+            Some(&new_genesis_data)
+        );
+    });
+}
+
+#[test]
+fn deregister_2_container_chains_in_same_block() {
+    new_test_ext().execute_with(|| {
+        run_to_block(1);
+        assert_ok!(ParaRegistrar::register(
+            RuntimeOrigin::signed(ALICE),
+            42.into(),
+            empty_genesis_data()
+        ));
+        assert_ok!(ParaRegistrar::register(
+            RuntimeOrigin::signed(ALICE),
+            43.into(),
+            empty_genesis_data()
+        ));
+        assert_ok!(ParaRegistrar::mark_valid_for_collating(
+            RuntimeOrigin::root(),
+            42.into(),
+        ));
+        assert_ok!(ParaRegistrar::mark_valid_for_collating(
+            RuntimeOrigin::root(),
+            43.into(),
+        ));
+
+        run_to_session(2);
+        assert_eq!(
+            ParaRegistrar::registered_para_ids(),
+            vec![42.into(), 43.into()]
+        );
+        assert_eq!(
+            ParaRegistrar::para_genesis_data(ParaId::from(42)).as_ref(),
+            Some(&empty_genesis_data())
+        );
+        assert_eq!(
+            ParaRegistrar::para_genesis_data(ParaId::from(43)).as_ref(),
+            Some(&empty_genesis_data())
+        );
+        assert_ok!(ParaRegistrar::deregister(RuntimeOrigin::root(), 42.into(),));
+        assert_ok!(ParaRegistrar::deregister(RuntimeOrigin::root(), 43.into(),));
+        assert_eq!(
+            Mock::get().called_hooks,
+            vec![
+                HookCall::Registered(42.into()),
+                HookCall::Registered(43.into()),
+            ]
```

Comment:
Maybe add a check that this hooks are called as part of the `mark_valid_for_collating` extrinsic.